### PR TITLE
[WIP] Support for fallbackLng (substituable language)

### PR DESCRIPTION
**Pull request by michalpulpan**
This PR tries to implement logic supporting `fallbackLng` object in the i18n config (more described in the #58). 
The requirement was basically to implement:
```javascript
{
   locales: ['fr', 'fr-FR', 'fr-BE', 'en', 'es', 'pt'],
   fallbackLng: {
    'fr-FR': ['fr'],
    'fr-BE': ['fr'],
   }
}
```
So that we don't have to store redundant data in our `_routes.json` and be fully compactible with [i18n-next fallbackLng](https://github.com/i18next/next-i18next/pull/1927).

It implements a function `getPathFromPaths` (sorry for this naming `getPath` is already used `plugin/getRouteBranchReRoutes.ts` 👀 ).
It was necessary to extend `types.ts` a bit with (`fallbackLng` object in `TNtrData`) and to create custom `NTRI18NConfig` extending original `I18NConfig`.

With this we should be now able to resolve the logic described above.

❗ But, this doesn't (yet) implement logic of [nonExplicitSupportedLngs](https://github.com/i18next/next-i18next/pull/1930) and it could still raise warning of [Max custom routes reached](https://nextjs.org/docs/messages/max-custom-routes-reached). I'm currently not 100% sure how to implement it so we could do different PR for it?.

All previous test cases pass and since we have to use different `defaultNtrData` in test cases (due to `fallbackLng` dict
) I added `setEnvDataFallbackLng` function with which we have to run before `fallbackLng` is tested (or it's presence). Hope it's clear and okay like this. It's a lot of redundant test but I did not workout how to make basically test it more intelligently without ruining previous tests (and adding `fallbackLng` everywhere).

Thank you for your feedback @cvolant !

### CODE REVIEW

**Review by cvolant — COMMENTED** [review]
Hi @michalpulpan, thank you for your PR.

It sounds good to me.
Concerning the redundant tests, you can leave it like this. One day perhaps, I may merge them with the others.

There is only 2 little things: see comments below.
Oh, and I don't understand why the yarn.lock has changed. Maybe we don't have the same node version or something.

If you want to implement nonExplicitSupportedLngs, it is probably better to do it in a new PR.

This PR is currently a draft: once you are ready, you can launch the example locally, then check that everything works. It it does, you can leave a comment to say it does, remove the "draft" state, then ask for a final review. I will then check it too, then merge it.

**Comment by cvolant** [comment]
Hi @michalpulpan! Any news on this?

File: src/plugin/getPathFromPaths.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  if (paths[locale]) {
+    return paths[locale] || paths.default
+  }
+  const ntrData = getNtrData()
+  if (!ntrData) {
+    return paths.default
+  }
+  const { fallbackLng } = ntrData
+
+  if (fallbackLng && fallbackLng?.[locale]) {
```
[the remark this answers is not in the document]
Comment (michalpulpan) [reply src/plugin/getPathFromPaths.ts:20]: ✅

File: tests/react/setEnvData.ts [hunk 2]
```diff
@@ -19,3 +29,10 @@ export const setEnvData = (ntrData: Partial<TNtrData> = {}) => {
     ...ntrData,
   }
 }
+
+export const setEnvDataFallbackLng = (ntrData: Partial<TNtrData> = {}) => {
+  window.__NEXT_TRANSLATE_ROUTES_DATA = {
+    ...defaultNtrDataFallbackLng,
+    ...ntrData,
+  }
+}
```
[the remark this answers is not in the document]
Comment (michalpulpan) [reply tests/react/setEnvData.ts:38]: Done, thank you. I had to tweak the object a bit 
```typescript
setEnvData({
      defaultLocale: 'fr-FR',
      locales: ['fr', 'fr-FR', 'fr-BE', 'en', 'es', 'pt'],
      fallbackLng: {
        'fr-FR': ['fr'],
        'fr-BE': ['fr'],
      },
    })
```
but it gets the job done 😊 .

**Comment by michalpulpan** [comment]
@cvolant  Pushed proposed fixes. 😊 Please let me know how to deal with the yarn.lock. Otherwise I'm ready!

**Comment by michalpulpan** [comment]
> Hi @michalpulpan, thank you for the update, and sorry for the delay, I was off for some days.
> 
> To solve the yarn.lock, you can try: `git checkout master -- yarn.lock`.
> 
> Then I will merge it and release a new version.

Hi, no problem 😊. Thank you for the trick, hopefully it worked out and `yarn.lock` is merged from master

**Pull request merged**